CORDA-2050 Upgrade Corda to Java 11 (compatibility mode)

josecoll opened the pull request:
## Summary
This PR contains the first phase migration of Corda to build and execute on a Java 11 VM.
The codebase maintains 100% source code compatibility with Java 8 (ie. changes do not use any of the new Java 9+ JDK libraries). This compatibility mode entails all code executes within the global `UNNAMED` module space (eg. the equivalent of the Java 8 global classpath), so there is no effective true module isolation per se.

We use the Gradle `sourceCompatibility` and `targetCompatibility` flags to specify compatibility modes, such that:
-  `sourceCompatibility = VERSION_1_8`
-  `targetCompatibility = VERSION_1_8` or `VERSION_11` (dependent on JDK used to execute gradle)

**Note**: a specific patched version of Java 11 is being used due to an officially reported OpenJDK bug related to cryptographic libraries used by Corda. The official OpenJDK 11 patch is expected on 15th October 2019.

Please see [JDK 11 Migration](https://r3-cev.atlassian.net/wiki/spaces/CCD/pages/1183909338/JDK+11+Migration) for further background and detailed migration steps completed.

## Dependencies
The Java 11 build of Corda requires:
- an upgraded version of Quasar built for Java 11.
  The version of Quasar to use may be driven from system properties (the default is to use the `quasarVersion` specified in `constants.properties`)
   For Java 11, these must be set as follows: `-DquasarVersion=0.8.0 -DquasarClassifier`
- Corda Gradle plugins upgraded to use aforementioned version of Quasar - see [CORDA-2924](https://r3-cev.atlassian.net/browse/CORDA-2924)
- Gradle 5: existing master codebase has already been upgraded to use this - see [CORDA-2893](https://r3-cev.atlassian.net/browse/CORDA-2893) (and an associated task to upgrade the capsule plugin to use Gradle 5: [CORDA-2895](https://r3-cev.atlassian.net/browse/CORDA-2895)).
- **a patched version of the official OpenJDK 11** (note: we initially patched the OpenJDK in-house, subsequently reported and obtained an official patch from our JDK supplier, and are now awaiting an official patch due on or around the 15th October 2019 in line with [Oracle's Critical Patch Update](https://www.oracle.com/technetwork/topics/security/alerts-086861.html) quarterly release). The OpenJDK reported bug in question can be found [here](https://bugs.openjdk.java.net/browse/JDK-8225745).

## Constraints
The Kotlin compiler does not support usage of Java 9 module directives (eg. to specify that private internal JDK packages may still be used in the global UNNAMED module space). It was necessary to re-write the `SharedMemoryIncremental` Kotlin class in Java to be able to specify the use of a private internal package at compile time as follows:
```
        if (JavaVersion.current() == JavaVersion.VERSION_11)
            options.compilerArgs = [
                    '--add-exports', 'java.base/sun.nio.ch=ALL-UNNAMED'
            ]
```

Corda Java 11 docker images included in this PR (eg. see `Dockerfile11`) temporarily use an in-house dockerized patched JDK 11 image published to an internal R3 docker registry.

## DJVM and SGX compatibility
The modules `core`, `core-deterministic`, `serialization`, `serialization-deterministic` strictly target Java 8 JVM bytecode to support deployment and execution of contract verification logic in an SGX hosted deterministic JVM.

## Testing
- All Unit, Integration and Smoke tests have been executed in a Continuous Integration environment (as per Corda master).
- Basic samples (Bank Of Corda) and compatibility tests (eg. a transacting Behave scenario combining Corda nodes built with Java 8 and 11).
- Basic performance testing using the R3 performance cluster (no specific Java 11 JVM / GC tuning applied)
- Specific tools and samples require additional testing.

## Outstanding issues
- Node Explorer does not currently work due to code requiring upgrade to latest stable version of  [TornadoFX](https://github.com/edvin/tornadofx/tree/jdk10) (and then switchover to a SNAPSHOT version with Java 10+ support). Work to complete this is being tracked here: https://r3-cev.atlassian.net/browse/CORDA-3208

## Further work
This PR is part of a larger Epic described [here](https://r3-cev.atlassian.net/browse/CID-725), and which includes future activities such as:
- true (Java 9) modularisation (inclusion of `module-info.java` and explicit imports/exports)
- adoption of new Java 11 JDK libraries and language features - [CORDA-2051](https://r3-cev.atlassian.net/browse/CORDA-2051)
- use JLink to create optimised runtime images - [CORDA-2754](https://r3-cev.atlassian.net/browse/CORDA-2754)
- improve startup and footprint using Application Class Data Sharing (AppCDS) - [CORDA-2755](https://r3-cev.atlassian.net/browse/CORDA-2755)

Commit:
Include Gradle JPMS plugin (v1.5.0)

Commit:
Add explicit reference to JAXB.

Commit:
Include JavaFX plugin (specify dependent JavaFX modules) and apply changes to relevant modules (explorer, demobench, client/jfx).

Commit:
TEMP exclude *-deterministic modules from build.

Commit:
SIMM valuation sample: do not use shrink custom task by default (and only use for JDK 1.8 due to Proguard version not supporting JDK 11)

Commit:
Remove illegal imports: sun.security, sun.reflect.

Commit:
TEMP disable use of "sun.misc.VMSupport" (re-instate with module add-exports directive)

Commit:
TEMP disable use of "sun.security, sun.misc.Unsafe" in serializer lists (re-instate with module add-exports directive)

Commit:
Change checkJavaVersion() startup check to support JDK 11.

Commit:
Add comment to Gradle JPMS plugin version.

Commit:
Revert: Add comment to Gradle JPMS plugin version.

Commit:
CORDA-2801: Test to check compatibility between TLS 1.2 and TLS 1.3 (#4993)

The test is currently disabled till we move to Java 11 (or beyond) when TLS 1.3 becomes available as part of JDK.
Local testing been performed with Open JDK 12 (12+33) and the test is passing.

Commit:
Update to use Corda Gradle plugins 5.0.0-SNAPSHOT

Commit:
Temp remove usage of java modularity plugin.

Commit:
Upgrade jacoco to JDK11 supported version

Commit:
Upgrade quasar to JDK11 supported version (0.8.0) -  dependent on Corda "quasar-utils" gradle plugin upgrade.

Commit:
Allow corda gradle plugin snapshot version resolution from artifactory 'corda-dev'

Commit:
Update IDE compiler dependencies to run tests within IntelliJ.

Commit:
Move test Java schemas to Kotlin as they are used only by Kotlin JUnit test (was causing ASM compilation failure).

Commit:
Move test Java schemas to Kotlin as they are used only by Kotlin JUnit test (was causing ASM compilation failure).

Commit:
Fix JUnit by adjusting assertion to reflect improved uncompressed byte size.

Commit:
Fixed JUnit to not use a deprecated/removed JDK package "com.sun.xml.internal.messaging.saaj".

Commit:
Update Java Version checking.

Commit:
Remove usage of private JDK class "sun.misc.Signal" (REVISIT)

Commit:
Included TLS 1.3 unit tests (see https://r3-cev.atlassian.net/browse/CORDA-2801)

Commit:
Remove usage of private JDK class "sun.security.rsa.RSAPrivateCrtKeyImpl" (REVISIT)

Commit:
Enable JDK-internal API illegal access warnings.

Commit:
Move Java unit test into kotlin package to prevent ASM compilation/classloading error (REVISIT)

Commit:
Move Java unit test into kotlin package to prevent ASM compilation/classloading error (REVISIT)

* What went wrong:
Execution failed for task ':core:test'.
> failed to read class file /Users/josecoll/IdeaProjects/corda-jdk11/core/build/classes/java/test/net/corda/core/flows/FlowsInJavaTest$PrimitiveReceiveFlow.class

Commit:
Move Java unit test into kotlin package to prevent ASM compilation/classloading error (REVISIT)

* What went wrong:
Caused by: org.gradle.api.GradleException: failed to read class file /Users/josecoll/IdeaProjects/corda-jdk11/node/build/classes/java/test/net/corda/node/services/events/FlowLogicRefFromJavaTest$JavaNoArgFlowLogic.class

Commit:
EXPERIMENTAL: tweaks and attempts to set module directives (with/without using gradle module plugin)

Commit:
EXPERIMENTAL: tweaks and attempts to set module directives (with/without using gradle module plugin)

Commit:
Enhanced JDK security debugging for JCA provider(s).
Used whilst investigating "Unrecognized algorithm for signature parameters SHA256withECDSA" JDK bug using Bouncy Castle.

Commit:
EXPERIMENTAL: tweaks and attempts to set module directives (with/without using gradle module plugin)

Commit:
Upgraded Mockito and targetCompatibility to 11 (REVISIT).

Commit:
Temporary remove Kotlin JUnit test that requires module directives to access private packages (sun.security.util, sun.security.x509)
Awaiting Kotlion compiler support: https://youtrack.jetbrains.com/issue/KT-20740

Commit:
Added configurable flag to continue on test failure so TC can perform complete test execution sweep.

Commit:
Remove invalid compiler flag (--illegal-access=warn is a run-time flag only).

Commit:
Remove usage of deprecated URLClassloader (re-coded without scanning and pattern matching on run-time classpath URLs)

Commit:
Fix quasar path for run-time agent instrumentation.

Commit:
Fix attempt to access boot classpath from runtimeMXBean (in JDK9+)

Commit:
Merge commit 'ba675e73b98f3276af27c235430922e2321b8763' into feature/jdk11-migration

Commit:
Bump Gradle plugins version to 5.0.1-SNAPSHOT to incorporate Quasar 0.8.0 upgrade

Commit:
Fix failing jar signing tests caused by JDK 11 keytool not supporting separate key password for PKCS12 KeyStores.

Commit:
Fix one more failing jar signing test caused by JDK 11 keytool not supporting separate key password for PKCS12 KeyStores.

Commit:
EXPERIMENTAL: Gradle test tweak to see if fixes TC reported failure:
java.lang.IllegalArgumentException: There is more than one CorDapp containing the package net.corda.finance.workflows on the classpath
Specify a package name which is unique to the CorDapp.

Commit:
Revert previous commit (didn't solve TC problems)

Commit:
TEMP revert introduction of JUnit 5 in :finance:workflows due to failing TC errors:
org.junit.vintage.engine.execution

Commit:
Repoint Corda gradle plugins to latest master snapshot + quasar 0.8.0 (JDK11 support)

Commit:
TEMP re-instate JUnit 5 declarations in :finance:workflows.

Commit:
Tuning :node:integrationTest with
    maxHeapSize = "1g"
    forkEvery = 10

Commit:
Fine tune :node tests: maxHeapSize = 1536m, forkEvery = 5

Commit:
Fix stricter bytecode verification trapped errors when carpentering Enums.
Fixes serialization tests throwing "java.lang.IncompatibleClassChangeError" upon dynamically invoking "valueOf" for an Enum.

Commit:
Increase JUnit forEvery parameter due to determine whether cause of Gradle exit failures:
Process 'Gradle Test Executor 1' finished with non-zero exit value 137

Commit:
Remove Gradle UseJUnitPlatform() until performance degradation issues resolved.

Commit:
Fix incorrect process exit.

Commit:
Revert maxHeapSize in :node as TC VM unable to allocate memory.

Commit:
Fine tuning of forkEvery to prevent TC agent failing with:

Execution failed for task ':node:test'. org.gradle.process.internal.ExecException: Process 'Gradle Test Executor 12' finished with non-zero exit value 137
This problem might be caused by incorrect test process configuration.
Please refer to the test execution section in the User Manual at https://docs.gradle.org/5.4.1/userguide/java_testing.html#sec:test_execution

Commit:
Fine tuning to prevent TC agent: java.lang.OutOfMemoryError: Java heap space

Commit:
Reset forkEvery=1 to resolve TC error:
    // java.lang.IllegalArgumentException: There is more than one CorDapp containing the package net.corda.finance.workflows on the classpath
    // Specify a package name which is unique to the CorDapp.

Commit:
Fix failing Serialization tests caused by StackTraceElement comparison failures (JPMS 9 introduces additional non-equal fields)

Commit:
Fixed failing Serialization tests.

Commit:
Fix failing smoke tests reporting:
java.lang.NoClassDefFoundError: com/esotericsoftware/kryo/Serializer

Commit:
Fix deployNodes gradle execution failure (caused by NetworkBootstrapper not finding Kryo classes at run-time).

Commit:
Resolve package split problem in Finance module causing:
There is more than one CorDapp containing the package net.corda.finance.workflows on the classpath
Problem caused by usage of 'net.corda.finance.Currencies' by both 'contracts' and 'workflows' modules.

Commit:
Fix dodgy IntelliJ package rename refactor (in smoke test code)

Commit:
Fix failed IntelliJ package rename refactor (in unit/integration test code)

Commit:
Revert "Fix dodgy IntelliJ package rename refactor (in smoke test code)"

This reverts commit ffeb6005

Commit:
Revert "Fix failed IntelliJ package rename refactor (in unit/integration test code)"

This reverts commit e1432cdf

Commit:
Fix "java.lang.IllegalArgumentException: There is more than one CorDapp containing the package net.corda.finance.workflows on the classpath" problems in finance module
caused by Jar loader processing *javadoc and *sources jars files.

Commit:
Revert "Resolve package split problem in Finance module causing: There is more than one CorDapp containing the package net.corda.finance.workflows on the classpath Problem caused by usage of 'net.corda.finance.Currencies' by both 'contracts' and 'workflows' modules."

This reverts commit bb7d6ff1

Commit:
Remove usage of JavaScript within log4j2.xml to prevent JVM attempt to load private scripting package:

Execution worker for ':' INFO Cannot initialize scripting support because this JRE does not support it. java.lang.NoClassDefFoundError: javax/script/ScriptEngineManager
at org.apache.logging.log4j.core.script.ScriptManager.<init>(ScriptManager.java:69)
caused by JavaScript embedded in log4j.xml (remove as not required)

Commit:
Fix invalid usage of URLClassloader to locate Jolokia JAR from classpath.

Commit:
Fix failing Jolokia integration test (switch from indirect agent based bootstrapping to direct Jolokia server bootstrapping).

Commit:
Fix failing SIMM Valuation demo by explicitly signing contracts JAR (without shrinking via ProGuard)

Commit:
Add explicit versioning to JFX modules.

Commit:
Merge branch 'master' into feature/jdk11-migration

Commit:
Upgrade ProGuard to JDK 11 supported version (plus reconfigure JDK system libraries).

Commit:
Minor cleanup to SIMM Valuation ProGuard gradle config.

Commit:
Re-instate deterministic module processing (core, serialization).
Adjust ProGuard configuration accordingly.
TODO: review failing "checkDeterminism" jar (predicated on a Java 8 RT).

shamsasari committed:
CORDA-2333: Dependency update pass for tests and demos

Commit:
Fix deterministic module issues by reverting to generating 1.8 target bytecode.
Note: this required disabling all "--add-exports" compiler directives (but at the same time moving inaccessible JDK library code sun.security.provider.SecureRandom from Kotlin to Java)

roastario committed:
add a shared memory port allocator to allow multiple processes to share a single allocation pool

roastario committed:
remove dangerous reset function on port allocator

roastario committed:
set forkCount = 2 in node integration test

roastario committed:
only allow one build of a cordapp at any given time for Driver tests

roastario committed:
make all portallocation requests use same starting point

roastario committed:
globally set forks to 6

roastario committed:
tweak forking parameters to allow parallel builds

roastario committed:
tweak unit test parallelism

roastario committed:
2 workers for integrationTest

roastario committed:
some more tweaks for parallel builds

roastario committed:
some more tweaks for parallel builds

roastario committed:
seems that 49K is not the start of ephemeral ports on all kernels

roastario committed:
tweak parallel settings

roastario committed:
try fix RPC shutdown test in parallel env

roastario committed:
add some logging for RPC shutdown test

roastario committed:
added some logging around PortAllocation tests - try figure out where they are getting stuck

roastario committed:
added some logging around PortAllocation tests - try figure out where they are getting stuck

Commit:
Upgrade log4j and slf4j dependencies to latest stable version.

Commit:
Commit failing deployNodes tasks by using simple slf4j logger.
Usage of log4j would require "--add-modules java.scripting" in Java 11 (see https://stackoverflow.com/questions/53714010/log4j2-slf4j-and-java-11)

Commit:
Merge branch 'master' into feature/jdk11-migration

Commit:
Temporarily @Ignore consistently hanging integration test.

chrisr3 committed:
CORDA-3015: Add StatePointer classes to corda-core-deterministic. (#5227)

Commit:
Minor test spawn tuning to attempt to prevent hanging FlowOverrideTests

shamsasari committed:
CORDA-2333: Reverting jersey and mockito as it currently causes issues with ENT (#5228)

r3domfox committed:
ENT-3444 create test-db module (#5093)

* ENT-3444 define RequiresDB annotation and junit5 extension

* Move to internal

* info to trace

* Limit exposure of gradle imports

* Enable annotation inheritance, and multiple SQL scripts per class or method

* Get the test context class globally for all groups

* usingRemoteDatabase flag

roastario committed:
fix api-scanner tests

shamsasari committed:
All uses of CheckpointStorage.getAllCheckpoints() close the stream after use (#5230)

shamsasari committed:
CORDA-2333: Updated the majority of the dependencies that were out of date

Some where not touched, including the crypto ones, libraries related to messaging and Kotlin.

shamsasari committed:
CORDA-2333: Updated the majority of the dependencies that were out of date

Some where not touched, including the crypto ones, libraries related to messaging and Kotlin.

# Conflicts:
#	.idea/compiler.xml
#	build.gradle
#	constants.properties

Commit:
Merge branch 'master' into feature/jdk11-migration

roastario committed:
minimize api changes

Commit:
Re-instate correct Mockito version for JDK11 following merge from master.

Commit:
Merge branch 'master' into feature/jdk11-migration

Commit:
Re-instate correct Mockito version for JDK11 following merge from master.

Commit:
Merge branch 'feature/jdk11-migration' of https://github.com/corda/corda into feature/jdk11-migration

roastario committed:
revert to complying with existing API

Commit:
Changes to enable JFX code to be cross-compilable between JDK 8 and 11

Commit:
Changes to enable build time selection of quasar dependency to be cross-compilable between JDK 8 and 11

Commit:
Revert usage of Java 9 API for version checking so code is cross-compilable between Java 8 and 11

Commit:
Changes to enable JFX code to be cross-compilable between JDK 8 and 11

roastario committed:
add the AtomicInteger for api compatibility reasons

Commit:
Remove references to org.javamodularity.moduleplugin to enable cross-compilation across JDK 8 and 11

Commit:
Remove JDK 11 compilation argument:  --illegal-access=warn

Commit:
Revert usage of Java 9 API for version checking so code is cross-compilable between Java 8 and 11

Commit:
Use `packageName` in a cross-compatible manner across JDK 8 and 11

roastario committed:
make sizing script executable

Commit:
Force timeout of tests prone to hanging.

roastario committed:
address review comments pt1

roastario committed:
address review comments pt2

roastario committed:
fix compile errors after review comments

roastario committed:
return to using home dir as temp dir seemed to interact badly with gradle

roastario committed:
Merge remote-tracking branch 'origin/master' into shared_memory_integer_allocation

roastario committed:
Merge remote-tracking branch 'origin/master' into shared_memory_integer_allocation

roastario committed:
Merge remote-tracking branch 'origin/master' into shared_memory_integer_allocation

Commit:
Merge branch 'master' into feature/jdk11-migration

Commit:
Minor fix after rebase from master.

Commit:
Merge branch 'master' into feature/jdk11-migration

Commit:
Revert Shared Memory port allocator using private JDK internal packages.
https://github.com/corda/corda/pull/5223

Commit:
Revert "add a shared memory port allocator to allow multiple processes to sha… (#5223)"

This reverts commit 88894bc5

Commit:
Fix merge from master conflicts.

Commit:
Fix 3 failing unit tests due to missing Kryo runtime dependency.

Commit:
Merge branch 'master' into feature/jdk11-migration

Commit:
Clean-up in preparation for PR.

Commit:
Dockerise Azul JDK 11 (patched) and create a Corda Node docker image using Azul JDK 11 (patched)

Commit:
Update release label so Behave framework can uniquely resolve Docker image from version label.

Commit:
Revert release label change.

Commit:
Publish Zulu 11 Docker images.

Commit:
Merge branch 'master' into feature/jdk11-migration

Commit:
Merge branch 'shared_memory_integer_allocation' into feature/jdk11-migration

Commit:
TEMP remove usage of VMSupport.

Commit:
Remove redundant references to quasar jdk8, kryo and kotlin libs.

Commit:
Configure correct target JVM bytecode (11) so we can use module directives to reference private, non-exported JDK packages.
Only "core" and "serialisation" require targeting Java 8 bytecode (DJVM/SGX constraint)

Commit:
Port SharedMemoryIncremental PortAllocation class to Java and add exports to reference private internal package: java.base/sun.nio.ch=ALL-UNNAMED

Commit:
Temporarily disable "targetCompatibility = VERSION_1_8" for serialization (as causing runtime resolution error upon gradle parsing)

Commit:
Fix compilation errors following rebase from master.

Commit:
Re-align with master (eg. @Ignore all TLS 1.3 interop tests).

Commit:
Fix failing unit test.

Commit:
@Ignore flaky RRPCStability test (the irony!).

Commit:
Re-written code using VMSupport to detect run-time agents.

Commit:
Set target compatibility based on executing java version.

Commit:
@Ignore flaky RRPCStability test (the irony!).

Commit:
Minor tweaks required for Java 8 compatibility.

Commit:
Merge branch 'master' into feature/jdk11-migration

Commit:
Remove redundant jvmArg declaration.

Commit:
Minor clean-up

Commit:
Re-instate use of sun.misc.Signal for disabling Ctrl-C

Commit:
Remove duplicate @Ignore for flaky tests.

Commit:
Default quasar version to Java 8 (so PR doesn't choke with standard GitHub TC gates).

Commit:
Adjust test assertion to cater for different checkpoint sizes according to quasar/kryo version in play (Java 8 / 11)

mikehearn commented:
171 commits .... should probably be squashed a bit! :o

mikehearn commented on build.gradle:
```diff
@@ -14,7 +17,13 @@ buildscript {
 
     ext.quasar_group = 'co.paralleluniverse'
-    ext.quasar_version = constants.getProperty("quasarVersion")
+    // Java 11: pass in -DquasarVersion=0.8.0 -DquasarClassifier
```
Can we set these automatically based on the version of Java, as seen below?

mikehearn commented on client/jfx/src/main/java/net/corda/client/jfx/utils/ReplayedList.java:
```diff
@@ -0,0 +1,16 @@
+import javafx.collections.ObservableList;
+import org.jetbrains.annotations.NotNull;
+
+// Java 9 introduces a new abstract method that we need to override (without using the explicit Kotlin `override` keyword to be backwards compatible
```
Hm, not sure I fully understand this. What's wrong with using the override keyword? Wouldn't work with Java 8?

I wonder if it's worth being compatible with both 8 and 11 simultaneously - that's the source of a lot of complexity here. What do we lose if we just take the leap and use git rollbacks to go to back 8 if we need to?

josecoll replied:
Kotlin forces you to specify the `override` keyword whilst Java doesn't.
Thus we can only make this cross-compatible via Java.

>>Re: wonder if it's worth being compatible with both 8 and 11 simultaneously

The intent is for this 1st stage migration to Java 11 to be 100% cross-compatible (single source, two targets: 8 & 11). Subsequent work with Java 11 will entail forking the code at some point.

mikehearn replied:
Yeah, I get the intent. I just wonder if it'd be worth shrinking this PR and just ... doing it. <insert nike tick here>

Wouldn't it be a bit confusing for users of the platform if we supported 8 and 11 simultaneously? Can app devs use 11 features or not?

josecoll replied:
That is an interesting question .... to which I think the answer is "maybe".
the Public API's (core module) and Serialization are still built to generate 100% Java 8 bytecode.
but that shouldn't stop CorDapps using JDK11 features right?
Time to adapt a sample standalone CorDapp to validate this.

mikehearn commented on client/rpc/build.gradle:
```diff
@@ -98,6 +98,9 @@ dependencies {
     smokeTestRuntimeOnly "org.junit.vintage:junit-vintage-engine:${junit_vintage_version}"
     smokeTestRuntimeOnly "org.junit.platform:junit-platform-launcher:${junit_platform_version}"
+
+    // JDK11: required by Quasar at run-time
```
Huh, why is Quasar getting involved in the RPC client code? We are only meant to use it on the server.

josecoll replied:
Agree - this is an item that has been raised on several occasions but never put to bed properly (I am guessing we have some entanglement in our module dependencies that doesn't make it easy to remove).
I will raise a separate JIRA to investigate / resolve this properly (independent of this Java 11 migration PR).

josecoll replied:
@blsemo would you care to share you thoughts (as Modularity WG lead) on how we might want to approach this?

mikehearn commented on node/src/main/kotlin/net/corda/node/serialization/kryo/DefaultKryoCustomizer.kt:
```diff
@@ -97,7 +95,6 @@ object DefaultKryoCustomizer {
             register(BufferedInputStream::class.java, InputStreamSerializer)
             register(Class.forName("sun.net.www.protocol.jar.JarURLConnection\$JarURLInputStream"), InputStreamSerializer)
             noReferencesWithin<WireTransaction>()
-            register(ECPublicKeyImpl::class.java, publicKeySerializer)
```
Hm, why are these removed?

josecoll replied:
Neither of these types are exported publicly:

import **sun.security.ec.ECPublicKeyImpl**:
```
module jdk.crypto.ec {
    provides java.security.Provider with sun.security.ec.SunEC;
}

```
import **sun.security.provider.certpath.X509CertPath**:

```
module java.base {
    ...
    exports sun.security.provider.certpath to java.naming;
    ...
}
```

Why are they being exposed for Serialization ?

mikehearn replied:
Good question. I'm not sure. Maybe @rick-r3 remembers?

rick-r3 replied:
Hmmm..... I wonder why we don't just register against the `PublicKey` and `PrivateKey` interfaces?

I'm fairly certain we could `addDefaultSerializer(PublicKey::class.java, publicKeySerializer)` and same for private.

mikehearn commented on node/src/main/kotlin/net/corda/node/utilities/JVMAgentUtil.kt:
```diff
@@ -0,0 +1,25 @@
+import java.lang.management.ManagementFactory
+import java.util.*
+
+object JVMAgentUtil {
```
I wonder if we really need this anymore.

josecoll replied:
This is a new utility used to provide a cross-compatible java 8 & 11 agent lookup mechanism.
See its usage in `NodeStartup` and `CheckpointDumper`.

mikehearn commented on samples/simm-valuation-demo/build.gradle:
```diff
@@ -31,14 +31,21 @@ dependencies {
+    cordapp project(':samples:simm-valuation-demo:contracts-states')
     cordapp project(':samples:simm-valuation-demo:flows')
 
+    // Cordformation needs a SLF4J implementation when executing the Network
```
Pretty grim that we're complicating the build file for apps even further. Developer experience is going down the toilet - this magic bit of boilerplate isn't documented anywhere. Is there no fix on the log4j side, or can we hide this away in the Gradle plugins somehow?

josecoll replied:
I believe @chrisr3 attempted to fix this. 
Perhaps he can shed some light on how to remove this unnecessary boilerplate?

mikehearn commented on samples/simm-valuation-demo/contracts-states/build.gradle:
```diff
@@ -8,7 +8,7 @@ cordapp {
     signing {
         // Cordapp is signed after the "shrink" task.
-        enabled false
+        enabled true
```
?

josecoll replied:
Will reset and test.

rick-r3 replied:
and?

josecoll replied:
Reset and re-tested SIMM Valuation demo. All worked fine.

josecoll replied:
OK - So the automated integration test is failing:

```
Caused by: net.corda.core.CordaRuntimeException: class "net.corda.vega.analytics.OGSIMMAnalyticsEngine"'s signer information does not match signer information of other classes in the same package
Caused by: net.corda.core.CordaRuntimeException: java.lang.SecurityException: class "net.corda.vega.analytics.OGSIMMAnalyticsEngine"'s signer information does not match signer information of other classes in the same package

net.corda.vega.SimmValuationTest > runs SIMM valuation demo FAILED
    java.io.IOException at SimmValuationTest.kt:82
```

Will troubleshoot further.

mikehearn commented on serialization/src/main/kotlin/net/corda/serialization/internal/carpenter/ClassCarpenter.kt:
```diff
@@ -346,7 +346,7 @@ class ClassCarpenterImpl @JvmOverloads constructor (override val whitelist: Clas
             visitLdcInsn(Type.getType("L${schema.jvmName};"))
             visitVarInsn(ALOAD, 0)
-            visitMethodInsn(INVOKESTATIC, jlEnum, "valueOf", "(L$jlClass;L$jlString;)L$jlEnum;", true)
+            visitMethodInsn(INVOKESTATIC, jlEnum, "valueOf", "(L$jlClass;L$jlString;)L$jlEnum;", false)
```
Wasn't this previously fixed to be correct? IIRC this flag is controlling whether it's an interface method or something like that, why is it changing here?

josecoll replied:
This fixes 4x Serialization unit tests failing with same root cause (ICCE):

```
:serialization:testnet.corda.serialization.internal.amqp.DeserializeNeedingCarpentryOfEnumsTest » singleEnum (0.016s)
java.lang.reflect.InvocationTargetException: (No message provided)Open stacktrace
Caused by: java.lang.IncompatibleClassChangeError: Method java.lang.Enum.valueOf(Ljava/lang/Class;Ljava/lang/String;)Ljava/lang/Enum; must be Methodref constant
```

Possible tightening of invokedynamic in JDK9:
http://hg.openjdk.java.net/jdk9/jdk9/hotspot/rev/ea45fb60fbbe

mikehearn reviewed (commented):
Wow, great work. Truly a mission.

How is packageName implemented in Java 11? I wonder if the packageName_ change should just be cleaned up so we use the real support in Java 11 (or package.name) and undo the nasty optimisations, knowing we'll get the speed back as part of the upgrade?

Commit:
Merge branch 'master' into feature/jdk11-migration

Commit:
Set quasar version according to Java version used to invoke gradle.

Commit:
Remove inclusion of Kotlin implementation of SharedMemoryIncremental (following merge from master).

Commit:
Fix building of Corda fat jar to include correct specification of quasar agent dependency.

Commit:
Fix failing Jar Signing unit tests (caused by behavioural difference in jarsigner tool between JDK 8 and 11).

Commit:
Merge branch 'release/os/4.3' into feature/jdk11-migration

rick-r3 commented on build.gradle:
```diff
@@ -201,9 +224,17 @@ allprojects {
+    if (JavaVersion.current() == JavaVersion.VERSION_1_8)
+        targetCompatibility = VERSION_1_8
+    else
+        targetCompatibility = VERSION_11
```
Do the artefacts produced in any (external) way reflect their target version? e.g. in filenames.  I'm guessing not.  Maybe it would be something we would do when publishing to Artifactory etc.  Could get confusing though.

josecoll replied:
> Do the artefacts produced in any (external) way reflect their target version? e.g. in filenames. I'm guessing not. Maybe it would be something we would do when publishing to Artifactory etc. Could get confusing though.

Yes - that is a good point.
Perhaps we should use the POM classifier to differentiate between Java 8 and 11 artifacts.

As per http://maven.apache.org/pom.html
```
classifier:
The classifier distinguishes artifacts that were built from the same POM but differ in content. It is some optional and arbitrary string that - if present - is appended to the artifact name just after the version number.
As a motivation for this element, consider for example a project that offers an artifact targeting JRE 1.5 but at the same time also an artifact that still supports JRE 1.4. The first artifact could be equipped with the classifier jdk15 and the second one with jdk14 such that clients can choose which one to use.

josecoll replied:
@rick-r3 as discussed, lets do this as a separate follow-up PR.
I've created a JIRA story for this: https://r3-cev.atlassian.net/browse/CORDA-3224

rick-r3 commented on config/dev/log4j2.xml:
```diff
@@ -22,41 +22,12 @@
-                            <PatternMatch key="WithMDC" pattern="%highlight{[%level{length=5}] %date{HH:mm:ssZ} [%t] %c{2}.%method - %msg %X%n%throwable{0}}{INFO=white,WARN=red,FATAL=bright red}"/>
-                        </ScriptPatternSelector>
-                    </PatternLayout>
+                    <PatternLayout pattern="%highlight{[%level{length=5}] %date{HH:mm:ssZ} [%t] %c{2}.%method - %msg%n%throwable{0}}{INFO=white,WARN=red,FATAL=bright red}"/>
```
I vaguely remember discussing this.  This is because Nashorn is deprecated, right ?  What about the bits of Javascript a few lines above this?  Is it completely removed / now an additional JAR or just not available?

josecoll replied:
Hmm ... this was one of those horrible run-time class visibility exception scenarios.
It manifested itself upon executing the _example-code:deployNodes_ gradle task:

```
[:example-code:deployNodes] 2019-06-05 00:28:33,803 Execution worker for ':' INFO Cannot initialize scripting support because this JRE does not support it. java.lang.NoClassDefFoundError: javax/script/ScriptEngineManager
[00:28:34]
[:example-code:deployNodes] 	at org.apache.logging.log4j.core.script.ScriptManager.<init>(ScriptManager.java:69)
```
See details in:
https://stackoverflow.com/questions/53714010/log4j2-slf4j-and-java-11

Both the above and https://issues.apache.org/jira/browse/LOG4J2-2488 suggest we need to execute the deploy command with `--add-modules java.scripting`. 
However, if you look at the module descriptor for `javax.script` you can see this package is already being exported:

```
module java.scripting {
    exports javax.script;

    uses javax.script.ScriptEngineFactory;
}
```

josecoll replied:
I have re-instated the original configuration as it no longer appears to be causing a problem (when just built locally). Lets see what a full TC build / test cycle thinks.

rick-r3 commented on constants.properties:
```diff
@@ -3,7 +3,7 @@
 
 cordaVersion=4.3-SNAPSHOT
-gradlePluginsVersion=5.0.2
+gradlePluginsVersion=5.0.1-SNAPSHOT-quasar-jdk11
```
Remind me what is going on here?

josecoll replied:
Aha! .... I need to update this to the latest gradle snapshot which should automatically handle the correct version of quasar according to the setting of `quasarVersion` and `quasarClassifier` in the main project gradle file.

josecoll replied:
Repointed to latest published version of Gradle Plugins (which contains quasar version and classifier parsing): https://ci-artifactory.corda.r3cev.com/artifactory/corda-releases/net/corda/plugins/quasar-utils/5.0.2/

rick-r3 commented on core-deterministic/testing/src/test/java/net/corda/deterministic/CheatingSecurityProvider.java:
```diff
@@ -0,0 +1,56 @@
+
+/**
+ * Temporarily restore Sun's [SecureRandom] provider.
+ * This is ONLY for allowing us to generate test data, e.g. signatures.
```
Is this in Java so it can use the extra exports to see the Sun classes?  Worth a comment here.

josecoll replied:
I have added similar comments elsewhere for Java re-written code.

rick-r3 commented on core/src/main/kotlin/net/corda/core/internal/CordaUtils.kt:
```diff
@@ -41,6 +41,7 @@ fun checkMinimumPlatformVersion(minimumPlatformVersion: Int, requiredMinPlatform
     }
 }
 
+// JDK11: revisit (JDK 9+ uses different numbering scheme: see https://docs.oracle.com/javase/9/docs/api/java/lang/Runtime.Version.html)
```
When do we need to revisit ?

josecoll replied:
I guess when we obtain and prescribe a specific version of Java 11.
The current patched Azul Zulu version we are using is:

```
colljos@ci-agent-101l:~$ java -version
openjdk version "11.0.3" 2019-04-16 LTS
OpenJDK Runtime Environment Zulu11.31+16-SA (build 11.0.3+7-LTS)
OpenJDK 64-Bit Server VM Zulu11.31+16-SA (build 11.0.3+7-LTS, mixed mode)
```

rick-r3 commented on core/src/test/kotlin/net/corda/core/internal/X509EdDSAEngineTest.kt:
```diff
@@ -1,125 +0,0 @@
-import net.i2p.crypto.eddsa.EdDSAPublicKey
-import org.junit.Test
-import sun.security.util.BitArray
-import sun.security.util.ObjectIdentifier
```
Could we restore this if we port to Java ?  Or did we decide these tests were duplicated elsewhere or something?

josecoll replied:
How bizarre .... I ported these tests to Java ... on the Enterprise feature branch (doh!!!).
Will add to this feature branch.

rick-r3 commented on node/src/test/kotlin/net/corda/node/services/network/DBNetworkParametersStorageTest.kt:
```diff
@@ -98,24 +96,11 @@ class DBNetworkParametersStorageTest {
-        block()
-        System.setOut(oldOut)
-        return out.bytes.inputStream().bufferedReader().lines().toList()
-    }
```
What doesn't work here?  Is `setOut` gone or something ?  Or is it the `ByteOutputStream` that looks like it was accidentally imported from the wrong place?

josecoll replied:
The test was using an internal class called: `com.sun.xml.internal.messaging.saaj.util.ByteOutputStream` to capture console output as a means of testing the correctness of an internal method (that should return an Exception but returns `null`).

rick-r3 commented on serialization-deterministic/build.gradle:
```diff
@@ -8,6 +10,9 @@ apply plugin: 'idea'
 evaluationDependsOn(":serialization")
 
+// required by DJVM and Avian JVM (for running inside the SGX enclave) which only supports Java 8.
+//targetCompatibility = VERSION_1_8
```
Is this supposed to be commented out?

josecoll replied:
see my comment below.

rick-r3 commented on serialization/build.gradle:
```diff
@@ -1,9 +1,14 @@
 description 'Corda serialization'
 
+// required by DJVM and Avian JVM (for running inside the SGX enclave) which only supports Java 8.
+//targetCompatibility = VERSION_1_8
```
What about this one?

josecoll replied:
OK - I can see what is happening here now.
This module has a _**testCompile**_ dependency on **_:node-driver_** (which is compiled to Jaava 11 bytecode), thus giving the following compile time exception:

```
* What went wrong:
Could not determine the dependencies of task ':serialization:test'.
> Could not resolve all task dependencies for configuration ':serialization:testRuntimeClasspath'.
   > Could not resolve project :test-common.
     Required by:
         project :serialization
      > Unable to find a matching variant of project :test-common:
          - Variant 'apiElements' capability net.corda:test-common:4.3-SNAPSHOT:
              - Incompatible attributes:
                  - Required org.gradle.jvm.version '8' and found incompatible value '11'.
                  - Required org.gradle.usage 'java-runtime' and found incompatible value 'java-api-jars'.
              - Other attributes:
                  - Found org.gradle.category 'library' but wasn't required.
                  - Required org.gradle.dependency.bundling 'external' and found compatible value 'external'.
                  - Required org.jetbrains.kotlin.platform.type 'jvm' and found compatible value 'jvm'.
          - Variant 'runtimeElements' capability net.corda:test-common:4.3-SNAPSHOT:
              - Incompatible attribute:
                  - Required org.gradle.jvm.version '8' and found incompatible value '11'.
              - Other attributes:
                  - Found org.gradle.category 'library' but wasn't required.
                  - Required org.gradle.dependency.bundling 'external' and found compatible value 'external'.
                  - Required org.gradle.usage 'java-runtime' and found compatible value 'java-runtime-jars'.
                  - Required org.jetbrains.kotlin.platform.type 'jvm' and found compatible value 'jvm'.
```
The solution to this is to split out the _serialization_ module tests into a separate module (as @blsemo has recently done for the _core_ module *and* using a different package name to avoid the split packages problem when we come to adding module descriptors to these modules).

josecoll replied:
I have created a separate JIRA to track this work (as it will be merged directly into the mainline OS and ENT branches): https://r3-cev.atlassian.net/browse/CORDA-3206

josecoll replied:
This is now completed and has been merged into this PR.

rick-r3 commented on testing/node-driver/src/main/java/net/corda/testing/driver/SharedMemoryIncremental.java:
```diff
@@ -0,0 +1,74 @@
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel;
+
+class SharedMemoryIncremental extends PortAllocation {
```
Worth a comment about it being in Java.

josecoll replied:
Yes - although @roastario mentioned to me the other day that they (the infrastructure team???) are ripping out this approach and reverting to a non-intrusive parallelisation mechanism (which I think is the correct way of solving the problem properly)

josecoll replied:
Added comment.

rick-r3 reviewed (commented):
First pass

Commit:
Merge branch 'release/os/4.3' into feature/jdk11-migration

Commit:
Re-instate MDCSelector (javascript code) as no longer appears to cause problems when built and run locally.

Commit:
Remove dependencies on javax.xml.bind (no longer used as per https://github.com/corda/corda/pull/5412)
CORDA-3175 Remove dependency on 3rd party javax.xml.bind library for simple hex parsing/printing.

Commit:
Add Java version of X509EdDSAEngineTest (which depends on internal JDK packages)

Commit:
Re-point to latest published version of Corda Gradle Plugins.

Commit:
Revert signing to original behaviour.

Commit:
Upgrade JFX	FontAwesomeFX dependencies to Java 11 supported.

Commit:
JFX: replace reference to old internal com.sun.javafx package to new javafx collection class.

shamsasari commented on webserver/src/main/java/CordaWebserverCaplet.java:
```diff
@@ -174,8 +174,8 @@ private void augmentClasspath(List<Path> classpath, File dir) {
         String version = System.getProperty("java.version");
-        if (version == null || !version.startsWith("1.8")) {
-            System.err.printf("Error: Unsupported Java version %s; currently only version 1.8 is supported.\n", version);
+        if (version == null || Arrays.asList("1.8", "11").stream().noneMatch(version::startsWith)) {
```
You can simplify this with `Stream.of("1.8", "11")`

Commit:
Incorporating code conciseness improvement from Shams.

Commit:
Incorporating RP suggestion: Register PublicKey and PrivateKey as default serializers.

Commit:
Fix SIMM valuation integration test (broken by not using the correct signed shrunk contracts jar)

Commit:
Add missing import java.util.stream.Stream

Commit:
Align jackson version to match constant declaration.

Commit:
Merge branch 'release/os/4.3' into feature/jdk11-migration

Commit:
Switch targetCompatibility = VERSION_1_8 back on for serialization module (following tests decoupling from node-api).

Commit:
Revert recent serialization change causing array deserialization failure in "KotlinIntegrationTestingTutorial.alice bob cash exchange example"

Commit:
Merge branch 'release/os/4.3' into feature/jdk11-migration

Commit:
Fix compilation warnings to comply with new code quality guidelines in CI (many introduced since Java 9).

Commit:
Re-instate usage of deprecated javax.security.cert.X509Certificate type (required by Artemis JAAS login).

Commit:
Added comments to Java code rewritten from Kotlin (to gain access to private internal JDK packages)

Commit:
Re-enable targetCompatibility = VERSION_1_8 in serialization-deterministic module.

Commit:
Delete code removed on master.

Commit:
Added changelog entry.

Commit:
Remove println: "Applying JFX plugin"

Commit:
Remove unused repository references (corda-dev, jpms module plugin)

Commit:
Revert timeout rules on Flow tests.

Commit:
Revert unnecessary fork per test.

Commit:
General clean-up in line with master.

Commit:
General clean-up in line with master.

rick-r3 reviewed (approved):
Can you work with @r3domfox team to create a PR gate for JDK 11 compatibility ?  I guess up to them whether they run per PR or we have something like the Windows build, but the latter isn't good at keeping things green.  Create a JIRA to track if we don't have one.

josecoll commented:
Yes - I created a JIRA to create a JDK11 compilation-only check gate: https://r3-cev.atlassian.net/browse/DEVOPS-1367
I propose we only run *tests* overnight as part of a snapshot build.
